web/handlers.go

```diff
@@ -157,8 +157,12 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 			c.Err.IsOAuth = false
 		}
 
-		w.WriteHeader(c.Err.StatusCode)
-		w.Write([]byte(c.Err.ToJson()))
+		if IsApiCall(r) {
```

Comment:
Jesús I don't think this will work as this endpoint will never be an API call, now the one thing we need to make sure is that if SAML returns an error and it does not have the action equals mobile like https://github.com/mattermost/mattermost-server/blob/master/web/saml.go#L90 then it should redirect, otherwise it should return the json object

Reply:
Ok @jespino as I said earlier this won't work.. I've been exploring options and have a solution basically we can check for the existence of a header like `X-Mobile-App` that if set then we return the json object if not then we redirect to the error page.

In case of the previous versions of the mobile app it will render the error page as it always have and with the new version of the app then it will render the error by parsing the json object.

basically your block would change to this:
```go
if len(r.Header.Get("X-Mobile-App")) > 0 {
    w.WriteHeader(c.Err.StatusCode)
    w.Write([]byte(c.Err.ToJson()))
} else {
    utils.RenderWebAppError(w, r, c.Err, c.App.AsymmetricSigningKey())
}
```

Reply:
I think we still need the `IsApiCall(r)` check because we want to return redirects only on non-api errors.

Reply:
I tested it manually, but I want to add unit tests.